packages/frames.js/src/core/middleware/renderResponse.tsx

```diff
@@ -0,0 +1,198 @@
+import { ImageResponse } from "@vercel/og";
+import { type Frame, getFrameFlattened, getFrameHtmlHead } from "../..";
+import type { ButtonProps } from "../components";
+import type { FrameDefinition, FramesMiddleware } from "../types";
+import { generatePostButtonTargetURL, isFrameRedirect } from "../utils";
+
+class InvalidButtonShapeError extends Error {}
+
+class InvalidButtonCountError extends Error {}
+
+class UnrecognizedButtonActionError extends Error {}
+
+class ImageRenderError extends Error {}
+
+/**
+ * This middleware is responsible for rendering the response
+ *
+ * If the accept header is set to application/json, it will return the metadata as JSON
+ * so it is easy to parse it for metatags in existing applications.
+ */
+export function renderResponse(): FramesMiddleware<{}> {
+  return async (context, next) => {
+    const wantsJSON =
+      context.request.headers.get("accept") === "application/json";
+
+    const result = await next(context);
+
+    if (result instanceof Response) {
+      return result;
+    }
+
+    if (isFrameRedirect(result)) {
+      return new Response(null, {
+        status: result.status ?? 302,
+        headers: {
+          ...result.headers,
+          Location: result.location.toString(),
+        },
+      });
+    }
+
+    try {
+      // @todo validate frame so it is according to spec and throw a ValidationError in case it isn't
+      // and handle that error in catch block
+      const frame: Frame = {
+        version: "vNext",
+        postUrl: "",
+        state: result.state ? JSON.stringify(result.state) : undefined,
```

Comment:
reminder that state cannot be included in the initial frame - we should check this and ensure that we are handling absent frame state in the initial POST request as well

Reply:
initial post request? you mean that on GET request the state should be thrown away right @stephancill ?

Reply:
The GET request state cannot have a state meta tag and therefore the initial POST request cannot expect state

Reply:
but there is no initial POST, only initial GET right? POST is only if you click button.

Reply:
there is an initial post in the sense that the first post request can never have state, but subsequent post requests can

Reply:
Ah I understand, since the GET response can't contain the state therefore when you click the button, there is no state. I think we just need to make sure, that `state` is not returned if the `request.method === 'GET'` right? And this also handles that when you click button the state is just not there. Is that good enough @stephancill?

Reply:
yes, makes sense

Reply:
done